File: c/tests/test_obd_dtc_find_0x6443E.c
/* test_obd_dtc_find_0x6443E.c
 *
 * Host C companion for obd_dtc_find_0x6443E (0x6443E).
 *
 * The DTC table (0xFFFF8930, 21 rows x 0x34) and the row-index word
 * (0xFFFF8D74) all live in page 0xFFFF8000, mapped MAP_FIXED.
 *
 * Reference: first i with byte@row+0x06 == (r4 & 0xFF) && i != currow
 *            returns s8(byte@row+0x08); else 0x08.
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>

extern int32_t obd_dtc_find_0x6443E(uint32_t r4);

#define BASE     0xFFFF8930u
#define STRIDE   0x34u
#define ROWS     0x15u
#define CURROW   0xFFFF8D74u

static void map_page(uintptr_t addr)
{
    long page = sysconf(_SC_PAGESIZE);
    uintptr_t base = addr & ~((uintptr_t)page - 1);
    void *p = mmap((void *)base, (size_t)page, PROT_READ | PROT_WRITE,
                   MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0);
    if (p == MAP_FAILED) {
        perror("mmap");
        exit(1);
    }
}

static uint8_t *rowp(uint32_t i) { return (uint8_t *)(BASE + i * STRIDE); }

static int32_t ref(uint32_t r4, uint16_t currow)
{
    for (uint32_t i = 0; i < ROWS; i++) {
        if (rowp(i)[0x06] == (uint8_t)r4 && i != currow)
            return (int32_t)(int8_t)rowp(i)[0x08];
    }
    return 0x08;
}

int main(void)
{
    unsigned failures = 0, tests = 0;

    map_page(0xFFFF8000);

    printf("=== obd_dtc_find_0x6443E ===\n");

    srand(42);
    for (int iter = 0; iter < 5000; iter++) {
        uint16_t currow = (uint16_t)(rand() & (ROWS - 1));
        for (uint32_t i = 0; i < ROWS; i++) {
            rowp(i)[0x06] = (uint8_t)rand();
            rowp(i)[0x08] = (uint8_t)rand();
        }
        *(volatile uint16_t *)CURROW = currow;
        uint32_t r4 = (uint32_t)(rand() & 0xFF);
        /* force a guaranteed hit every 3rd iteration */
        if (iter % 3 == 0) {
            uint32_t i = (uint32_t)(rand() % ROWS);
            if (i != currow) {
                rowp(i)[0x06] = (uint8_t)r4;
            } else {
                rowp((i + 1) % ROWS)[0x06] = (uint8_t)r4;
            }
        }
        tests++;
        int32_t got = obd_dtc_find_0x6443E(r4);
        int32_t exp = ref(r4, currow);
        if (got != exp) {
            printf("FAIL: iter=%d r4=%02X currow=%u got=%d expected=%d\n",
                   iter, r4, currow, got, exp);
            failures++;
            break;
        }
    }

    printf("Results: %u tests, %u failures\n", tests, failures);
    return failures ? 1 : 0;
}

File: c/tests/test_obd_dtc_find_0x6443E.py
#!/usr/bin/env python3
"""
Verify obd_dtc_find_0x6443E (0x6443E) against the ACTUAL ROM bytes, run
in the SH-2E emulator.

Search leaf, takes r4 (byte key).  Seeds all 21 DTC table rows' bytes 0x06
and 0x08 plus the active row index, then compares the ROM's return:

  for i in 0..0x14:
    p = 0xFFFF8930 + i * 0x34
    if byte@p+0x06 == (r4 & 0xFF) and i != word@0xFFFF8D74:
        return s8(byte@p+0x08)
  return 0x08

C:
  int32_t obd_dtc_find_0x6443E(uint32_t r4)

Run from repo root:  python3 c/tests/test_obd_dtc_find_0x6443E.py [N]
"""
import os, sys, random

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(ROOT, 'tools'))
from sh2emu import SH2

ROM = os.path.join(ROOT, 'roms', 'stock', '60E1D400.bin')
ENTRY = 0x006443E

BASE = 0xFFFF8930
STRIDE = 0x34
ROWS = 0x15
CURROW = 0xFFFF8D74


def s8(x):
    x &= 0xFF
    return x - 256 if x & 0x80 else x


def addr(row, off):
    return (BASE + row * STRIDE + off) & 0xFFFFFFFF


def main():
    N = int(sys.argv[1]) if len(sys.argv) > 1 else 2000
    rom = open(ROM, 'rb').read()
    cpu = SH2(rom)

    def run_one(r4, b06s, b08s, currow):
        ram = {CURROW: (currow >> 8) & 0xFF, CURROW + 1: currow & 0xFF}
        for i in range(ROWS):
            ram[addr(i, 0x06)] = b06s[i]
            ram[addr(i, 0x08)] = b08s[i]
        cpu.call(ENTRY, r4=r4, ram=ram)
        return cpu.r[0] & 0xFFFFFFFF

    def model(r4, b06s, b08s, currow):
        key = r4 & 0xFF
        for i in range(ROWS):
            if (b06s[i] & 0xFF) == key and i != currow:
                return s8(b08s[i]) & 0xFFFFFFFF
        return 0x08

    # Targeted: single-row hit at every index, plus current-row skip.
    for i in range(ROWS):
        b06s = [0] * ROWS
        b08s = [0] * ROWS
        b06s[i] = 0x5A
        b08s[i] = 0xC3  # s8 -> -61
        got = run_one(0x5A, b06s, b08s, (i + 1) % ROWS)
        exp = model(0x5A, b06s, b08s, (i + 1) % ROWS)
        if got != exp:
            print("FAIL: hit row %d -> 0x%X expected 0x%X" % (i, got, exp)); sys.exit(1)
        got = run_one(0x5A, b06s, b08s, i)
        exp = model(0x5A, b06s, b08s, i)
        if got != exp:
            print("FAIL: row %d current-row skip -> 0x%X expected 0x%X" % (i, got, exp)); sys.exit(1)

    # Random.
    for _ in range(N):
        r4 = random.randint(0, 0xFF)
        currow = random.randint(0, 0x14)
        b06s = [random.randint(0, 255) for _ in range(ROWS)]
        b08s = [random.randint(0, 255) for _ in range(ROWS)]
        got = run_one(r4, b06s, b08s, currow)
        exp = model(r4, b06s, b08s, currow)
        if got != exp:
            print("FAIL: r4=%02X currow=%d -> 0x%X expected 0x%X" % (r4, currow, got, exp))
            sys.exit(1)

    print("OK  obd_dtc_find_0x6443E @0x%04X  (targeted + %d random)" % (ENTRY, N))
    sys.exit(0)


if __name__ == '__main__':
    main()
